### ISSUE TITLE
Perform memory check for bulk read/write APIs

### ISSUE THREAD

**Issue by vkuznet** [dmwm/WMArchive#63]
After yesterday meeting I was working with a code to implement bulk read/write
functionality for WMArchive. It is doable, but I realized that based on current
structure of FWJR and Hadoop "desires" to deal with large files we may end-up
with a bottleneck on our server. Here is a problem:
- Kate/Zbigniew said that they prefer (Hadoop prefer) to write file of large
  size, the number they mention was about 256MB.
- In order to write this file we need to load/parse N documents
- Each FWJR document we deal with is JSON, moreover it has deep nested structure,
  i.e. dict contains list of dicts, etc.
- The Avro schema is also JSON and it order to write a bulk of docs we need
  to create a JSON like {'data': [our_fwjr_docs_here]} (*), we can't write just
  list of fwjr's since avro schema can't consume list of dics, it requires
  a name for the record [1](I think internally it is static data struct
  defined by schema and such struct should have named attributes).
- Let's assume that we want to create large avro file with N documents where N
  is large (to accommodate Hadoop desire). Then we'll end-up in python code
  which will hold N FWJR in memory, then we'll need to create (*) structure
  which will increase RAM usage. This is the same effect as I dealt with
  DBS memory leaks and the same as Marco reported for CRAB server.

I didn't make any estimates yet, since I don't know N, but we may have a
problem here. On one hand we can't hold too many docs due to potential
RAM usage, on another Hadoop "desire" large docs.

So, to make reasonable estimate I need to know N.

From WMArchive twiki I read the following:
"We are currently running 10 WMAgents, each agent has on average about 1M jobs,
each job will produce FWJR report which we need to store in WMArchive. All jobs
are done in parallel, but we expect to have on the order of <1K docs per
seconds feed into WMArchive. The system should sustain such load."

But from this sentence I don't know what is a daily rate we need to deal with.
Seangchan, can you provide better estimate. I need to know how many docs
each agent will write on average per day. Once we know that I can try different
N's, e.g. 1K, 10K, 100K and measure RAM usage when we'll load data from
MongoDB into code which will convert them to Avro file as a chunk.

Then I can have better idea which threshold for N we can sustain on our server.
So far, mongo2hdfs script reads docs from Mongo and write them (one-by-one) to
HDFS. To support bulk write, I need to read N docs from Mongo, keep them in RAM
in order to write Avro file to HDFS.  Of course I can split last step and write
N docs into local file system and then run separate process to copy that file
into HDFS, but it still does not solve RAM usage issue.

But the real issue here is a FWJR data-structure we use and python memory
allocation for it. Since FWJR has deep nestness structure and we need
to hold large chunk of them again in a dictionary I can't organize streaming.
Yet, it is another issue with FWJR and storage/look-up approach.
The storage and look-up prefer flat data, while our application relies
on deep nested dict structures. The real remedy is to modify FWJR structure
to be flat and then use streaming for writing. The data look-up will benefit
from flat structure as well, e.g. compare a.b.c.d.e.f data access with just
single column look-up.

Best,
Valentin.

[1] http://docs.oracle.com/cd/E26161_02/html/GettingStartedGuide/avroschemas.html#avro-creatingschemas

**Comment by vkuznet** [comment]
here some numbers I got with the following exercise.
I read FWJR processing document, 25KB in size, replicated it to N docs,
injected into MongoDB and read it back from MongoDB by creating bulk doc
which I discussed in my email, i.e. {'data':[fwjr dics here]}

For all operations I measured RAM usage using psutil. I manually checked that
psutil reports correct info by comparing its output with output of top
(I found deviation about few %, but that's probably due to different
implementation of psutil and top tools).

Here is breakdown:
- injecting 1000 docs into Mongo took 25MB
- reading 1000 docs from Mongo and creating bulk doc with them: ~100MB
- injecting 10K docs into Mongo took 30MB
- reading 10K docs from Mongo and creating bulk doc with them: ~750MB

(note: psutil reports different numbers every time I run command [1], but I
think this is due to the fact I run on live system which has other processes).

There are two observations. The injection RAM usage is small, because I used
generator and injected data in chunks with 1k chunk size.
The reading is a bottleneck as I expected due to python memory allocation
and the fact that we need to allocate memory for all docs and create bulk record.

I also measured size of bulk document when I write it as JSON to local file
system. It is 112MB for bulk doc with 10K records in it. If I take 25 as
a compression factor from JSON to gzipped avro, it means that bulk doc
will have size 4.5MB, but it will cost us ~750MB of RAM on server side.

The commands I run like this (code is here [1]):
./t.py --fin=test/data/fwjr_processing.json --ndocs=10000
./t.py --action=read

That means that we probably can afford 10K docs to be written at once to HDFS,
it will cost us about 750MB of RAM on server and produced gziped avro file will
be 4.5MB.

I'll put this observation to ticket [2] for future references.

Best,
Valentin.

[1] https://gist.github.com/vkuznet/2dee7724172dca36624d
[2] https://github.com/dmwm/WMArchive/issues/63
